### DISCUSSION TITLE
Best way for a test to reference which function it is testing?

### DISCUSSION THREAD

**Discussion by JustinGrote** [pester/Pester#1984] [Ideas]
So I've been playing with the new vscode proposed test api, and one neat thing about their markdown example is that the tests can get a codelens right next to the actual code, not just where the test exists, which is nice not having to bounce back and forth, especially when using the autorun.

![image](https://user-images.githubusercontent.com/15258962/121574042-bdd77680-c9da-11eb-8244-379bd358d186.png)

This got me thinking, maybe there should be a way for a test to reference what function it is testing (which would be completely optional). As a simple test I just used the describe block to match to the function name using an AST parser.

Should there maybe be an explicit parameter for describe/function/it blocks (e.g. "-For" to reference it is for a particular function) to codify this? It would just be an additional property on the block objects, nothing that pester would initially use directly except maybe to say Invoke-Pester -For 'Get-MyFunction'). 

@nohwnd thoughts?

**Comment by JustinGrote** [comment]
Related, if it was possible to get test results emitted as they complete as output objects I could make the test runner more responsive by updating the UI as the tests complete rather than waiting till the end. I suppose I could grok the text output but having a structured method would be much better. A pesterconfig output option that would be something like "objectsonly" that would return individual test objects as they complete and then the full output at the end.

**Reply by nohwnd** [reply]
I talked with @TylerLeonhardt a while ago about why is he not streaming the results to the UI (I know you are now talking about a different UI probably, the native one, but from I understood it is similar, but ported into VSCode directly), and he told me that the options of getting the results back into the extension are limited. You can definitely setup a netpipe, or similar IPC channel, and push the results in there as soon as Pester is done running the test. He is not doing that because there is not enough time. But imho that is the best option. But if you are able to get the output streaming directly from powershell then outputting the test objects is really simple. Just add a plugin in Pester and choose the step where it should run. Look at output.ps1 it has the plugin that writes to screen. There is also one for coverage and mock if you need more inspiration.

**Reply by JustinGrote** [reply]
Right, I'm talking [about the new proposed native interface](https://code.visualstudio.com/api/extension-guides/testing). It operates a bit different and has some more flexibility yes.

If I just do a stream exec of pwsh instead of using Powershell Editor services then the json should get emitted to stdout as they output, so it should allow for real time updates.

**Reply by nohwnd** [reply]
Then you just need to solve the problem of mapping the object from run to the one from discovery. https://github.com/TylerLeonhardt/vscode-pester-test-adapter/issues/30#issuecomment-859278023

**Reply by JustinGrote** [reply]
I should be able to reasonably do this now with the `NoRun`option shouldn't I? Doesn't `NoRun` "expand" the tests to the same object and description as they would be if it was actually run, including foreach/testcase runs?

**Reply by fflaten** [reply]
SkipRun will give you the right amount of test-obects, but ExpandedName + ExpandedPath aren't expanded until run-phase which is skipped so they will still include stuff like "\<varName\>". AFAIK there's no reliable unique identitfier for parameterized tests except their index/placement in the blocks.

**Reply by nohwnd** [reply]
Yup that's right. There is no link. It is a problem if tests are generated from changing data (say the number of seconds right now, to make a simple example). 

https://github.com/TylerLeonhardt/vscode-pester-test-adapter/issues/13#issuecomment-859270910

**Comment by nohwnd** [comment]
There is long time proposal from Gael to have metadata on tests. I recommended he tries using single item -ForEach. There you can specify any data, even data you don't use in your test. He told me that he will try it, but got no feedback. All these data are then on .Data on the result object. So I would recommend you start with that as well, before we try adding something like .Metadata.

**Reply by gaelcolas** [reply]
So I did look at -Foreach and it works well for my use case.
It's a bit too loosely typed for my exact need, so I need to wrap around and provide more options.

I'm interested in the output "plugin" you speak of and how to author one, but that's probably for later. I'll keep digging.
